tardis/utilities/asyncbulkcall.py

```diff
@@ -0,0 +1,173 @@
+    Single-task calls are buffered for a moment according to ``size`` and ``delay``,
+    then executed in bulk with ``concurrent`` calls to ``command``.
+
+    Each :py:class:`~.BulkExecution` should represent a different ``command``
+    (for example, ``rm`` or ``mkdir``) collecting similar tasks (for example,
+    ``rm foo`` and ``rm bar`` to ``rm foo bar``). The ``command`` is an arbitrary
+    async callable and can freely decide how to handle its tasks. The
+    :py:class:`~.BulkExecution` takes care of collecting individual tasks,
+    partitioning them to bulks, and translating the results of bulk execution
+    back to individual tasks.
+
+    Both ``size`` and ``delay`` control how long to queue tasks at most
+    before starting to execute them. The ``concurrent`` parameter controls whether
+    and how many bulks may run at once; when concurrency is low or disabled, tasks
+    may be waiting for execution even past ``size`` and ``delay``.
+    Possible values for ``concurrent`` are
+    :py:data:`False` (no concurrency, only one ``command`` at once),
+    either of :py:data:`True` or :py:data:`None` (unlimited concurrency),
+    or an integer above 0 to set a precise concurrency limit.
+
+    .. note::
+
+        If the ``command`` requires additional arguments,
+        wrap it via :py:func:`~functools.partial`.
+    """
+
+    def __init__(
+        self,
+        command: BulkCommand[T, R],
+        size: int,
+        delay: float,
+        concurrent: Union[int, bool, None] = True,
+    ):
+        self._command = command
+        self._size = size
+        self._delay = delay
+        self._concurrency = (
+            1
+            if concurrent is False
+            else sys.maxsize
+            if concurrent is None or concurrent is True
+            else concurrent
+        )
+        # task handling dispatch from queue to command execution
+        self._master_worker: Optional[asyncio.Task] = None
+        self._verify_settings()
+
+    @cached_property
+    def _concurrent(self) -> "asyncio.BoundedSemaphore":
+        """synchronized counter for active commands"""
+        return asyncio.BoundedSemaphore(value=self._concurrency)
+
+    @cached_property
+    def _queue(self) -> "asyncio.Queue[Tuple[T, asyncio.Future[R]]]":
+        """queue of outstanding tasks"""
+        return asyncio.Queue()
+
+    def _verify_settings(self):
+        if not isinstance(self._size, int) or self._size <= 0:
+            raise ValueError(f"expected 'size' > 0, got {self._size!r} instead")
+        if self._delay <= 0:
+            raise ValueError(f"expected 'delay' > 0, got {self._delay!r} instead")
+        if not isinstance(self._concurrency, int) or self._concurrency <= 0:
+            raise ValueError(
+                "'concurrent' must be one of True, False, None or an integer above 0"
+                f", got {self._concurrency!r} instead"
+            )
```

Comment:
This type check is not required when being executed from within `__init__` as you are taking care explicitly that the type does match. So I recommend simplifying it.

```suggestion
        if or self._concurrency <= 0:
            raise ValueError(
                "expected 'concurrent' > 0, got {self._concurrency!r} instead"
            )
```

Reply:
There's only a type hint in ``__init__``, but I never actually check if it's indeed an ``int``. I just know it's neither ``bool`` nor ``None``.